File: src/VibeApp.API/Shared/Common/Result.cs
namespace VibeApp.API.Shared.Common;

public enum ErrorType
{
    None,
    NotFound,
    Validation,
    Unauthorized,
    Forbidden,
    ServerError
}

public class Result
{
    public bool IsSuccess { get; }
    public string? Error { get; }
    public ErrorType ErrorType { get; }

    protected Result(bool isSuccess, string? error, ErrorType errorType = ErrorType.None)
    {
        IsSuccess = isSuccess;
        Error = error;
        ErrorType = errorType;
    }

    public static Result Success() => new(true, null);

    public static Result Failure(string error, ErrorType errorType = ErrorType.ServerError)
        => new(false, error, errorType);

    public static Result<T> Success<T>(T value) => new(value, true, null);

    public static Result<T> Failure<T>(string error, ErrorType errorType = ErrorType.ServerError)
        => new(default, false, error, errorType);
}

public class Result<T> : Result
{
    public T? Value { get; }

    protected internal Result(T? value, bool isSuccess, string? error, ErrorType errorType = ErrorType.None)
        : base(isSuccess, error, errorType)
    {
        Value = value;
    }
}

File: backend/app/shared/models/result.py
"""Result pattern for consistent API responses."""
from typing import Generic, TypeVar, Optional
from enum import Enum
from pydantic import BaseModel


class ErrorType(str, Enum):
    """Types of errors that can occur."""
    NOT_FOUND = "not_found"
    VALIDATION = "validation"
    UNAUTHORIZED = "unauthorized"
    FORBIDDEN = "forbidden"
    SERVER_ERROR = "server_error"


T = TypeVar('T')


class Result(BaseModel, Generic[T]):
    """Result wrapper for API responses."""
    is_success: bool
    value: Optional[T] = None
    error: Optional[str] = None
    error_type: Optional[ErrorType] = None

    @classmethod
    def success(cls, value: T) -> "Result[T]":
        """Create a successful result."""
        return cls(is_success=True, value=value)

    @classmethod
    def failure(cls, error: str, error_type: ErrorType = ErrorType.SERVER_ERROR) -> "Result[T]":
        """Create a failed result."""
        return cls(is_success=False, error=error, error_type=error_type)

    class Config:
        use_enum_values = True
